### ISSUE TITLE
cannot reverse proxy in front of grpc + grpc-web multiplexer

### ISSUE THREAD

**Issue by tiero** [caddyserver/caddy#3997]
I have a go server that does multiplexing of grpc and grpc.web based on content-type, this means I accept both http/1.1 and http2 with TLS. Unfortunately it seems impossible to have all version in the `http` transport directive of `reverse_proxy`, either I support plain text grpc-web or http2 binary grpc.

The go server expect the TLS cert and key (wich I generated with caddy)

## This is my Caddyfile

```
https://api.mydomain.com

log {
    level INFO
}

reverse_proxy {
    to h2c://api:8000
    transport http {
        versions h2c 2
    }
}
```

## This is what it logs caddy when calling /Service/Method  with a grpc client

```sh

{
  "level": "error",
  "ts": 1611876892.228595,
  "logger": "http.log.error.log0",
  "msg": "write tcp 172.30.0.3:53400->172.30.0.2:8000: write: broken pipe",
  "request": {
    "remote_addr": "5.90.3.181:36177",
    "proto": "HTTP/2.0",
    "method": "POST",
    "host": "api.mydomain.com",
    "uri": "/Service/Method",
    "headers": {
      "User-Agent": [
        "grpc-node/1.24.2 grpc-c/8.0.0 (linux; chttp2; ganges)"
      ],
      "Grpc-Accept-Encoding": [
        "identity,deflate,gzip"
      ],
      "Accept-Encoding": [
        "identity,gzip"
      ],
      "Te": [
        "trailers"
      ],
      "Content-Type": [
        "application/grpc"
      ]
    },
    "tls": {
      "resumed": false,
      "version": 771,
      "cipher_suite": 49196,
      "proto": "h2",
      "proto_mutual": true,
      "server_name": "api.mydomain.com"
    }
  },
  "duration": 0.005654583,
  "status": 502,
  "err_id": "n9kacgca2",
  "err_trace": "reverseproxy.statusError (reverseproxy.go:783)"
}
```

```sh
{
  "level": "error",
  "ts": 1611876892.2293694,
  "logger": "http.log.access.log0",
  "msg": "handled request",
  "request": {
    "remote_addr": "5.90.3.181:36177",
    "proto": "HTTP/2.0",
    "method": "POST",
    "host": "api.mydomain.com",
    "uri": "/Service/Method",
    "headers": {
      "Content-Type": [
        "application/grpc"
      ],
      "User-Agent": [
        "grpc-node/1.24.2 grpc-c/8.0.0 (linux; chttp2; ganges)"
      ],
      "Grpc-Accept-Encoding": [
        "identity,deflate,gzip"
      ],
      "Accept-Encoding": [
        "identity,gzip"
      ],
      "Te": [
        "trailers"
      ]
    },
    "tls": {
      "resumed": false,
      "version": 771,
      "cipher_suite": 49196,
      "proto": "h2",
      "proto_mutual": true,
      "server_name": "api.mydomain.com"
    }
  },
  "common_log": "5.90.3.181 - - [28/Jan/2021:23:34:52 +0000] \"POST /Service/Method HTTP/2.0\" 502 0",
  "duration": 0.005654583,
  "size": 0,
  "status": 502,
  "resp_headers": {
    "Server": [
      "Caddy"
    ]
  }
}
```

## My docker compose

```yaml
version: "3.7"

services:
  api:
    container_name: "my_container"
    image: my_api:latest
    restart: unless-stopped
    environment:
      - SSL_CERT=./cert.pem
      - SSL_KEY=./key.pem
    ports:
      - "8000:8000"
    volumes:
      - path/to/cert.crt:/cert.pem
      - path/to/key.key:/key.pem
  caddy:
    container_name: "caddy"
    image: caddy:latest
    restart: unless-stopped
    depends_on:
      - api
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - $HOME/Caddyfile:/etc/caddy/Caddyfile
      - $HOME/caddy-data:/data

```

**Comment by mholt** [comment]
I'm not sure I understand; you want to serve both HTTP and HTTPS on the same socket?

**Comment by tiero** [comment]
I'm serving on the same interface (host:port) both gprc and grpc-web with TLS termination. Meaning that I already have a multiplexer that handle this in my go code, namely I support requests coming from client as `application-grpc` and `application-grpc-web`/`application/grpc-web-text`

This works fine If I call my server directly. Not sure if this multiplexing is problem, just told as extra detail.

If I put Caddy in front as reverse proxy I cannot make it work.



This is my mux code if necessary

```

func serveMux(address string, withSsl bool, grpcServer *g.Server) error {
	lis, err := net.Listen("tcp", address)
	if err != nil {
		return err
	}

	if sslKey := config.GetString(config.SSLKeyPathKey); sslKey != "" && withSsl {
		certificate, err := tls.LoadX509KeyPair(config.GetString(config.SSLCertPathKey), sslKey)
		if err != nil {
			return err
		}

		const requiredCipher = tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
		config := &tls.Config{
			CipherSuites: []uint16{requiredCipher},
			NextProtos:   []string{"http/1.1", http2.NextProtoTLS, "h2-14"}, // h2-14 is just for compatibility. will be eventually removed.
			Certificates: []tls.Certificate{certificate},
		}
		config.Rand = rand.Reader

		lis = tls.NewListener(lis, config)
	}

	mux := cmux.New(lis)
	grpcL := mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldPrefixSendSettings("content-type", "application/grpc"))
	httpL := mux.Match(cmux.HTTP1Fast())

	grpcWebServer := grpcweb.WrapServer(
		grpcServer,
		grpcweb.WithCorsForRegisteredEndpointsOnly(false),
		grpcweb.WithOriginFunc(func(origin string) bool { return true }),
	)

	go grpcServer.Serve(grpcL)
	go http.Serve(httpL, http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
		if isValidRequest(req) {
			grpcWebServer.ServeHTTP(resp, req)
		}
	}))

	go mux.Serve()
	return nil
}
```

**Comment by jonnywei** [comment]
@tiero  you should change request header Accept-Encoding = identity ,see caddy source code [https://github.com/caddyserver/caddy/blob/e2940c8c03818f1227aa2c103e05fb5664635e11/modules/caddyhttp/reverseproxy/streaming.go#L126](url)

**Comment by tiero** [comment]
@jonnywei Do you mean to add in my client metadata? I'm afraid this could be a no-go if I need to asks client  (whihc are not run by myself) to add specifics header in their requests.

Could i manipulate this myself somehow in the caddyfile?

**Comment by jbrown-stripe** [comment]
@tiero Running into the exact same issue, any luck figuring it out?

**Comment by mholt** [comment]
@jbrown-stripe Let me know if I can be of any help here. grpc isn't my specialty, but if needed I could invest some time next week to reproduce this and investigate together.

**Comment by tiero** [comment]
> @tiero Running into the exact same issue, any luck figuring it out?

No unfortunately. I ended up implementing TLS via certmagic in gocode directly

But would be nice to use caddyserver anyway, bur cannot figure out still

**Comment by jbrown-stripe** [comment]
> > @tiero Running into the exact same issue, any luck figuring it out?
> 
> No unfortunately. I ended up implementing TLS via certmagic in gocode directly
> 
> But would be nice to use caddyserver anyway, bur cannot figure out still

@tiero 
I never needed http2 with tls, just h2c. Regardless, caddy is sending h2c requests when proxying `grpc-web`. However your http listener does not match it and cannot handle the request. It also doesn't like h2c request can be handled by the default http server in `go`. My solution was to
1. set `flush_interval=1` in caddy. This seemed to fix the stream errors I was running into. 
2. Explicitly handle h2c and http 1.1 requests within go
```
cMux := cmux.New(lis)
	grpcListener := cMux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
	grpcWebHTTP2Listener := cMux.Match(cmux.HTTP2())
	grpcWebHTTP1Listener := cMux.Match(cmux.HTTP1())

	wrappedGrpc := grpcweb.WrapServer(s.grpc)
	handleGrpcWebReq := func(w http.ResponseWriter, req *http.Request) {
		if wrappedGrpc.IsGrpcWebRequest(req) {
			wrappedGrpc.ServeHttp(w, req)
			return
		}

		msg := "received a request that could not be matched to grpc or grpc-web"
		log.Error(msg)
		w.WriteHeader(http.StatusNotFound)
		w.Write([]byte(msg))
	}

	mux := http.NewServeMux()
	mux.HandleFunc("/", handleGrpcWebReq)

	httpSrv := &http.Server{
		Addr:    s.address,
		Handler: mux,
	}

	http2Srv := &http.Server{
		Addr:    s.address,
		Handler: h2c.NewHandler(http.HandlerFunc(handleGrpcWebReq), &http2.Server{}),
	}

	serveGRPC(s.grpc, grpcListener)
	serveHTTP(http2Srv, grpcWebHTTP2Listener)
	serveHTTP(httpSrv, grpcWebHTTP1Listener)
```

```
func serveHTTP(s *http.Server, l net.Listener) {
	go func() {
		if err := s.Serve(l); err != nil {
			log.Fatal("error starting http server", log.Err(err))
		}
	}()
}

func serveGRPC(s *grpc.Server, l net.Listener) {
	go func() {
		if err := s.Serve(l); err != nil {
			log.Fatal("error starting grpc server", log.Err(err))
		}
	}()
}
```

**Comment by jbrown-stripe** [comment]
@tiero It looks like both grpc-web and grpc request get sent to the grpc listener which won't work. Instead only grpc requests should be routed to the grpc listener and then grpc-web should be routed to http1.1 and h2c listener.

**Comment by mholt** [comment]
@jbrown-stripe Can you help me understand, then, what is your Caddy reverse_proxy config now? And did you have to change your grpc backend (or client? :thinking:) to make it work? (I'm surprised that much needed to change to proxy through Caddy. It should be easier.)

**Comment by jbrown-stripe** [comment]
@mholt The caddy reverse proxy configuration is practically the same except I had set `flush_interval` to `-1`. 
```
{
    "handler": "reverse_proxy",
    "flush_interval": -1,
                  "transport": {
                    "protocol": "http",
                    "versions": [ "1.1", "h2c"]
                  },
}
```
I think the main issue stems from the fact that `grpc-web` requests are actually incompatible with a grpc server because the [protocols](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md) are slightly different. The general solution is to have a reverse proxy like Envoy sit between the client and the grpc server and I'm not sure Caddy supports translating the requests like Envoy. My solution for this was to use [grpcweb](https://github.com/improbable-eng/grpc-web/tree/master/go/grpcweb) to create a `grpc-web` compatible server and use a connection multiplexer to host both `grpc-web` and `grpc` on the same port. This is where I ran into issues, and I suspect this is also where @tiero had problems.

Caddy was only sending `h2c` requests to my webserver so while my `grpc` handler could understand it, my grpc-web handler could not and I was getting `read: connection reset by peer` errors. At first I thought this was a Caddy issue, but it looks to be an issue with my Go server. Since the default http server in Go does not handle `h2c` I had to modify the multiplexer to 
1. Match any http2 request with `application/grpc` headers with the grpc server
2. Match any h2c request with `application/grpc-web` headers with a http2 server
3. Match any http1.1 request with `application/grpc-web` headers with the default http server

TLDR: This really an issue with Caddy but an issue with how we wrote our gRPC servers. Caddy does not translate `grpc-web` to grpc so our Go servers need to handle grpc-web requests. Additionally, we have to explicitly use the h2 package `grpc-web` requests since the default go http server does not serve h2 by default.

**Comment by mholt** [comment]
Thanks for the explanation, I think that makes more sense to me now. So perhaps a handler or transport module that could translate between `grpc-web` and `grpc` would be useful.

Anyway, I think your explanation will help a lot of future searchers on the subject, so thank you!

**Comment by francislavoie** [comment]
@jbrown-stripe might be a longshot, but could you try building from the latest from `master`? 

I think there's a chance https://github.com/caddyserver/caddy/commit/e6c29ce081673d85e527d59f3afb7ace034573df has fixed this issue, but I may not be fully understanding the problem. Just a sanity check to make sure this is still an issue.

**Comment by tiero** [comment]
Hey @jbrown-stripe we refactored the server multiplexer and we used your same Caddy configuration, still we cannot route gRPC via caddy as reverse proxy, just grpc-web actually.


Ref: https://github.com/tdex-network/tdex-daemon/pull/423


Can you confirm that it works for you?

**Comment by kostyay** [comment]
I'm still getting this error with Caddy 2.4.6

My configuration is as follows:
```
{
        email xxx
}

xxx {
        encode zstd gzip
        log {
                level DEBUG
                output file /var/log/caddy/cli.log
        }

        reverse_proxy {
                to localhost:9090
                transport http {
                        tls_insecure_skip_verify
                        versions h2c 1.1
                }
        }
}
```

**Comment by tiero** [comment]
@kostyay same thing.

I suspect main problem is that we should be able to have a new directive that can switch transport based on content-type (the trick we do in out cmux based multiplexer) cc/ @mholt @francislavoie

**Comment by francislavoie** [comment]
@tiero that's an interesting idea. But you can just use the `header` request matcher to handle requests differently based on content type -- use a different `reverse_proxy` altogether.

**Comment by francislavoie** [comment]
I'll close this as inactive. I'm not seeing anything actionable in this issue at this point.

If someone comes up with a better explanation of the problem (with the latest version of Caddy) and can show that there's still an issue, then we can look into it again.

**Comment by tetra12** [comment]
I have the same issue in July 2022.
OS: Ubuntu 21.10
Caddy: 2.5.2-alpine and 2.4.6-alpine

I have a local GRPC service without TLS. I can ping it with `grpcurl` like this:
```bash
> grpcurl -plaintext -vv -d @ 192.168.1.93:50051 cc_service.CloudControlService/UnaryCall <ssh.json

Resolved method descriptor:
rpc UnaryCall ( .cc_service.APIRequest ) returns ( .cc_service.APIResponse );
Request metadata to send:
(empty)
Response headers received:
content-type: application/grpc

Estimated response size: 62 bytes

Response contents:
{
  "jobID": "cd083052-fca0-494d-ba1e-803bbc5cc337",
  "data": "eyJjb3VudCI6MCwiaXRlbXMiOltdfQ=="
}

Response trailers received:
(empty)
Sent 1 request and received 1 response
```


Then, I'm trying to have `Caddy` as a reverse proxy for my `grpc-web` requests

For `Caddy 2.5.2-alpine`, my GRPC requests are hanging - no data in console and a `pending` status in browser.

For `Caddy 2.4.6-alpine`,  I'm getting the following:
```bash
proxy    | {"level":"debug","ts":1657813382.5313122,"logger":"http.handlers.reverse_proxy","msg":"upstream roundtrip","upstream":"192.168.1.93:50051","duratio
n":0.001241764,"request":{"remote_addr":"192.168.176.1:45924","proto":"HTTP/1.1","method":"OPTIONS","host":"localhost:5052","uri":"/cc_service.CloudControlSer
vice/UnaryCall","headers":{"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Sec-Fet
ch-Mode":["cors"],"Sec-Fetch-Site":["same-site"],"Sec-Fetch-Dest":["empty"],"Referer":["http://localhost:9000/"],"Access-Control-Request-Method":["POST"],"Acc
ept-Encoding":["gzip, deflate, br"],"Accept-Language":["en-US,en;q=0.9"],"X-Forwarded-For":["192.168.176.1"],"X-Forwarded-Proto":["http"],"Accept":["*/*"],"Ac
cess-Control-Request-Headers":["content-type,x-grpc-web"],"Origin":["http://localhost:9000"]}},
"error":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x00\\x00\\x06\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x05\\x00\\x00@\\x00\""}


proxy    | {"level":"error","ts":1657813382.532908,"logger":"http.log.error","msg":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"
\\x00\\x00\\x06\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x05\\x00\\x00@\\x00\"","request":{"remote_addr":"192.168.176.1:45924","proto":"HTTP/1.1","method":"OPTIONS
","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTM
L, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Sec-Fetch-Mode":["cors"],"Sec-Fetch-Site":["same-site"],"Sec-Fetch-Dest":["empty"],"Referer":["http://loc
alhost:9000/"],"Connection":["keep-alive"],"Access-Control-Request-Method":["POST"],"Accept-Encoding":["gzip, deflate, br"],"Accept-Language":["en-US,en;q=0.9
"],"Accept":["*/*"],"Access-Control-Request-Headers":["content-type,x-grpc-web"],"Origin":["http://localhost:9000"]}},"duration":0.003359381,"status":502,"err
_id":"p0cejyrib","err_trace":"reverseproxy.statusError (reverseproxy.go:886)"}

```

My **Caddyfile**:
```bash
{
  debug
  # auto_https disable_certs

  servers {
    protocol {
      # We need to enable h2c in order to use grpc without TLS.
      # ref: https://caddy.community/t/caddy-grpc-h2c-passthrough/11780
      allow_h2c
    }
  }
}

http://localhost:5052 {

  reverse_proxy h2c://192.168.1.93:50051 {

    transport http {
      versions h2c 2
      dial_timeout 3s
    }
  }
}
```

**Comment by mholt** [comment]
@tetra12 Is your client speaking grpc-web and your backend speaking grpc? From the above comments, it sounds like those are not the same. Are you expecting Caddy to convert between the two?

**Comment by mholt** [comment]
Oh, I see. Thanks for clarifying.,

I thought this was a bug report, I didn't realize it was a feature request.

So if the feature request is a grpc-web bridge -- without knowing exactly what to do -- I guess we can do that. I don't understand how grpc-web works exactly, but if someone else has experience with it, then a pull request would certainly be welcomed and would also be a faster way to a solution than me figuring it out :sweat_smile: 

> How do I config Caddy to deliver that HTTP 1.1 to HTTP 2.0 transport bridging?

I'm a bit confused though; and forgive me, for not being familiar with the world of gRPC: Caddy can already translate between HTTP versions in its proxy because it terminates and decodes HTTP requests, then re-encodes new ones for the backend. So it can receive HTTP/1.1 and use HTTP/2 to the upstream. This should already work. Is that all the bridge is/does?

**Comment by tetra12** [comment]
@mholt Could you pls provide me links/examples to working configs? I searched both community site and issues here and I have NOT found a solution, which would work for GRPC for me

**Comment by tetra12** [comment]
My Caddyfile looks like this :eyes: 

```bash
{
  debug
  servers {
    protocol {
      # We need to enable h2c in order to use grpc without TLS.
      # ref: https://caddy.community/t/caddy-grpc-h2c-passthrough/11780
      allow_h2c
    }
  }
}

http://localhost:5052 {

  reverse_proxy h2c://192.168.1.93:50051 {
    transport http {
      versions h2c 2
      dial_timeout 3s
    }
  }
}
```

**Comment by mholt** [comment]
@tetra12 I believe that's what your Caddyfile does: you're serving on HTTP (not HTTPS), so that uses HTTP/1.1, and then `reverse_proxy h2c://...` tells Caddy to use HTTP/2 over cleartext to the backend. (I think your `versions h2c and 2` line is unnecessary here, but @francislavoie could double-check me on that.)

I'm not sure what your backend is replying with. This byte sequence isn't familiar to me, it definitely doesn't look like TLS or HTTP: `\\x00\\x00\\x06\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x05\\x00\\x00@\\x00\`

**Comment by francislavoie** [comment]
Yeah, all this is redundant and can be removed:

```
    transport http {
      versions h2c 2
      dial_timeout 3s
    }
```

The dial timeout as of v2.5.0 (iirc) is 3s, and using `h2c://` sets those versions automatically

**Comment by tetra12** [comment]
Okay, I was quietly hoping to avoid a mess with TLS for dev purposes :roll_eyes: . 
Long story short, for a self-generated local certs with `mkcert` I'm getting this :disappointed: :
```bash
TLS handshake error from 192.168.192.1:55084: remote error: tls: unknown certificate
```


Okay dokey, let's try this setup:

```bash
{
  debug
  # auto_https disable_certs

  servers {
    protocol {
      # We need to enable h2c in order to use grpc without TLS.
      # ref: https://caddy.community/t/caddy-grpc-h2c-passthrough/11780
      allow_h2c
    }
  }
}

https://localhost:5052 {

  reverse_proxy h2c://192.168.1.93:50051 {
    transport http {
      dial_timeout 5s
    }
  }
}
```

Then I generated a pair of certs with `mkcert` and installed a local CA:
```bash
> mkcert -install #installs a local CA
> mkcert localhost  127.0.0.1 0.0.0.0 192.168.64.1 192.168.1.93
```

When I installed those certs both on my backend and for `Caddy`, I'm getting this:
```bash
proxy    | {"level":"info","ts":1657835473.50363,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
proxy    | {"level":"info","ts":1657835473.5037339,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
proxy    | {"level":"info","ts":1657835473.50497,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0001ec7e0"}
proxy    | {"level":"info","ts":1657835473.5055046,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/data/caddy"}
proxy    | {"level":"info","ts":1657835473.5078769,"logger":"tls","msg":"finished cleaning storage units"}
proxy    | {"level":"info","ts":1657835473.5302165,"logger":"pki.ca.local","msg":"root certificate is already trusted by system","path":"storage:pki/authorities/local/root.crt"}
proxy    | {"level":"debug","ts":1657835473.5303,"logger":"http","msg":"starting server loop","address":"[::]:5052","http3":false,"tls":true}
proxy    | {"level":"debug","ts":1657835473.5303268,"logger":"http","msg":"starting server loop","address":"[::]:80","http3":false,"tls":false}
proxy    | {"level":"info","ts":1657835473.5303311,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["localhost"]}
proxy    | {"level":"warn","ts":1657835473.5325403,"logger":"tls","msg":"stapling OCSP","error":"no OCSP stapling for [localhost]: no OCSP server specified in certificate","identifiers":["localhost"]}
proxy    | {"level":"debug","ts":1657835473.532553,"logger":"tls.cache","msg":"added certificate to cache","subjects":["localhost"],"expiration":1657878311,"managed":true,"issuer_key":"local","hash":"d5a657a0ec08e953720236f6b331500d44424439a8b0fd0b688ae4b3373e4d03","cache_size":1,"cache_capacity":10000}
proxy    | {"level":"info","ts":1657835473.5328555,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
proxy    | {"level":"info","ts":1657835473.5328627,"msg":"serving initial configuration"}
proxy    | {"level":"debug","ts":1657835492.8590171,"logger":"tls.handshake","msg":"choosing certificate","identifier":"localhost","num_choices":1}
proxy    | {"level":"debug","ts":1657835492.8590415,"logger":"tls.handshake","msg":"default certificate selection results","identifier":"localhost","subjects":["localhost"],"managed":true,"issuer_key":"local","hash":"d5a657a0ec08e953720236f6b331500d44424439a8b0fd0b688ae4b3373e4d03"}
proxy    | {"level":"debug","ts":1657835492.8590465,"logger":"tls.handshake","msg":"matched certificate in cache","subjects":["localhost"],"managed":true,"expiration":1657878311,"hash":"d5a657a0ec08e953720236f6b331500d44424439a8b0fd0b688ae4b3373e4d03"}
proxy    | {"level":"debug","ts":1657835492.8605769,"logger":"http.stdlib","msg":"http: TLS handshake error from 192.168.192.1:55084: remote error: tls: unknown certificate"}
```

@mholt @francislavoie  Any ideas ? :thinking:

**Comment by mholt** [comment]
@tetra12 Ok, I have a few questions because I don't really understand what you're trying to do now.

Why are you enabling h2c at the same time as HTTPS? H2C doesn't get used with TLS.

If your goal is to enable TLS to the client:

```

localhost:5052 {
  reverse_proxy h2c://192.168.1.93:50051 {
    transport http {
      dial_timeout 5s
    }
  }
}
```

is all you need. No mkcert. No allow_h2c.

(Why are you using `mkcert`? Caddy generates its own certificates for you, and even installs its trust into your system's trust store. Unlike mkcert, Caddy even keeps these certificates renewed.)

Anyway, I'm interested in addressing this as I'm currently neck-deep in proxy stuff right now, so it'd be a good time for this (cognitively speaking). Thanks for helping us get to the bottom of this; but I think you were closer with your initial config.

If you don't need/want to use TLS with your client right now, then just use allow_h2c and leave HTTPS disabled (i.e. just add allow_h2c to your earlier config).

**Comment by tetra12** [comment]
@mholt Thanks for giving a great feedback. I was hoping to use `Caddy` like `Envoy` to convert `grpc-web` into `grpc` right away. But as @mohammed90 said, present Caddy impl lacks codecs and filters, which are impl in `Envoy`. 

`grpc` config would be more challenging due to filters, cors and extra params. You may look :eyes: into Envoy example [here](https://github.com/grpc/grpc-web/blob/7fe18d1e088c650002209141cfaaad96cc8949b6/net/grpc/gateway/examples/echo/envoy.yaml#L37)

As for now I'll try to follow @mohammed90  advice and deploy BOTH `Caddy` and `Envoy`. This is NOT an elegant solution of course, but may work.

I'll dig :eyes:  into `Caddy` code to see if GRPC could be impl straight away. It's a good idea to lift some code right from `Envoy` since it's also impl in `Go`.

Anyways, thanks for your support. If you wanna have some feedback from me, play around with your ideas or settings on proxy configuration, just let me know :sunglasses: 

If you have team discussions on `Caddy` dev, I'd love to join

And thanks all for building and maintaining `Caddy`. It's a new bright star in the ocean of traffic routing and I'd definitely navigate with it to discover new horizons :boat:

**Comment by mholt** [comment]
@tetra12 

> present Caddy impl lacks codecs and filters, which are impl in Envoy. ... grpc config would be more challenging due to filters, cors and extra params.

This isn't really accurate. "Codecs" and "filters" (though they have different names in this project) are very common in Caddy, even though our configs look different from Envoy's. We're quite capable and familiar with advanced proxy features like "codecs" (encodings/protocols) and "filters" (routing and middleware), and grpc would be no more "challenging" than websockets, NTLM, H2C, and other similar features Caddy already supports.

> You may look eyes into Envoy example [here](https://github.com/grpc/grpc-web/blob/7fe18d1e088c650002209141cfaaad96cc8949b6/net/grpc/gateway/examples/echo/envoy.yaml#L37)

Thanks for the links. I've seen a number of Envoy examples and am familiar with them by now, but seeing the config isn't really productive at this point -- it just needs to be implemented into Caddy, and there's not much Envoy can help with here because...

> It's a good idea to lift some code right from Envoy since it's also impl in Go.

Envoy is written in C++, actually. Not Go.

**Comment by tetra12** [comment]
@mholt The reason to use `mkcert` is to allow TLS on the backend. I need all services to talk natively with `mTLS` to each other.
I had a project with `Isio`, which has an automatic cert management in control plane with `istio-csr`. 

For Caddy, there's a lack of a higher level control plane which would allow sharing certificates between different services. That's why I need manually generate certs and manage them. Later in production, I may again turn to Service Mash to avoid pains with manual cert management :roll_eyes:

**Comment by mohammed90** [comment]
> @mholt The reason to use `mkcert` is to allow TLS on the backend. I need all services to talk natively with `mTLS` to each other. I had a project with `Istio`, which has an automatic cert management in control plane with `istio-csr`.
> 
> For Caddy, there's a lack of a higher level control plane which would allow sharing certificates between different services. That's why I need to manually generate certs and manage them. Later in production, I may again turn to Service Mesh to avoid pains with the manual cert management 🙄

You can use a shared storage backend besides the default of filesystem to share the certs. For instance, there are two module to utilize Consul as the certificate storage, from which you can grab the certs to be shared across instances.

**Comment by mholt** [comment]
@mohammed90 True, although I wonder if he means across different programs, i.e. not just Caddy. It's true other programs can use those certs too, but they have to know to look for them there.

**Comment by tetra12** [comment]
@mohammed90 @mholt 

What you say makes a good sense. Most microservice based systems would deploy Caddy as an edge proxy to do 3 things:
1. Automatic HTTPS on frontend(s)
2. Static data sharing
3. Direct traffic routing from internet to some internal services (e.g oauth2)

`Caddy` is great in doing those 3 things simply, quick and scalable.

However, all internal services in prod need to support `mTLS` nowadays. That implies 3 other things:
1. Each service needs an independent cert lifetime management
2. Internal traffic GRPC and REST won't come thru Caddy. Services would communicate directly or by external routing components ( Kubernetes routing or sidecar proxy in Meshes ).
3. You can't use a single cert managed by Caddy to run the whole system. Why? Because most prod deployments would want to allow some kind of RBAC. E.g a backend server would allow to communicate only with some specific services. 

That said, I'd love to use Caddy to serve its purpose and let cert mgmt, RBAC and other security concerns to be implemented by dedicated components.



Thus trying to use Caddy for cert management is just a misuse and won't play for us.

**Comment by tetra12** [comment]
Okay, let's back :back: to business :muscle: 

Here's my test for `caddy-2.5.2-alpine`:

```bash
{
  debug
  # auto_https disable_certs

  servers {
    protocol {
      # We need to enable h2c in order to use grpc without TLS.
      # ref: https://caddy.community/t/caddy-grpc-h2c-passthrough/11780
      allow_h2c
    }
  }
}

http://localhost:5052 { 

  reverse_proxy h2c://192.168.1.93:50051 {
  # reverse_proxy grpc://192.168.1.93:50051 {
    transport http {
      dial_timeout 5s
    }
  }
}
```
For that setup, I'm getting the following for my `grpc-web` request from the browser: `"error":"stream error: stream ID 3; PROTOCOL_ERROR; received from peer"}`

The full log:

```bash
proxy    | {"level":"debug","ts":1657872316.3449636,"logger":"http.handlers.reverse_proxy","msg":"selected upstream","dial":"192.168.1.93:50051","total_upstreams":1}
proxy    | {"level":"debug","ts":1657872381.4191055,"logger":"http.handlers.reverse_proxy","msg":"upstream roundtrip","upstream":"192.168.1.93:50051","duration":65.073252775,"request":{"remote_ip":"192.168.240.1","remote_port":"52896","proto":"HTTP/1.1","method":"OPTIONS","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"X-Forwarded-Host":["localhost:5052"],"Origin":["http://localhost:9000"],"Sec-Fetch-Site":["same-site"],"Accept-Encoding":["gzip, deflate, br"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Referer":["http://localhost:9000/"],"Sec-Fetch-Dest":["empty"],"X-Forwarded-For":["192.168.240.1"],"X-Forwarded-Proto":["http"],"Access-Control-Request-Method":["POST"],"Access-Control-Request-Headers":["content-type,x-grpc-web"],"Accept":["*/*"],"Sec-Fetch-Mode":["cors"],"Accept-Language":["en-US,en;q=0.9"]}},"error":"stream error: stream ID 3; PROTOCOL_ERROR; received from peer"}
proxy    | {"level":"error","ts":1657872381.4214551,"logger":"http.log.error","msg":"stream error: stream ID 3; PROTOCOL_ERROR; received from peer","request":{"remote_ip":"192.168.240.1","remote_port":"52896","proto":"HTTP/1.1","method":"OPTIONS","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"Access-Control-Request-Method":["POST"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Referer":["http://localhost:9000/"],"Connection":["keep-alive"],"Access-Control-Request-Headers":["content-type,x-grpc-web"],"Sec-Fetch-Dest":["empty"],"Accept-Language":["en-US,en;q=0.9"],"Accept":["*/*"],"Origin":["http://localhost:9000"],"Sec-Fetch-Mode":["cors"],"Sec-Fetch-Site":["same-site"],"Accept-Encoding":["gzip, deflate, br"]}},"duration":65.077869786,"status":502,"err_id":"k49v4h5rh","err_trace":"reverseproxy.statusError (reverseproxy.go:1184)"}
```

And the Web browser gives me this: 

```bash
account:1 Access to fetch at 'http://localhost:5052/cc_service.CloudControlService/UnaryCall' from origin 'http://localhost:9000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

Now I'm adding a CORS stuff, which worked for my REST APIs in a web browser:

```bash
{
  debug
  # auto_https disable_certs

  servers {
    protocol {
      # We need to enable h2c in order to use grpc without TLS.
      # ref: https://caddy.community/t/caddy-grpc-h2c-passthrough/11780
      allow_h2c
    }
  }
}

http://localhost:5052 {

  @options {
		method OPTIONS
	}

  header {
    Access-Control-Allow-Origin *
    Access-Control-Allow-Credentials true
    Access-Control-Allow-Methods *
    Access-Control-Allow-Headers *
    defer
  }

  reverse_proxy h2c://192.168.1.93:50051 {
    transport http {
      dial_timeout 5s
    }
  }

  respond @options 204
}

```

And I'm getting a protocol error again: `"http2: Transport: cannot retry err [stream error: stream ID 3; PROTOCOL_ERROR; received from peer] after Request.Body was written; define Request.GetBody to avoid this error"}`

Here's the full log:
```bash
proxy    | {"level":"debug","ts":1657872668.2121153,"logger":"http.handlers.reverse_proxy","msg":"selected upstream","dial":"192.168.1.93:50051","total_upstreams":1}
proxy    | {"level":"debug","ts":1657872668.2152174,"logger":"http.handlers.reverse_proxy","msg":"upstream roundtrip","upstream":"192.168.1.93:50051","duration":0.002510926,"request":{"remote_ip":"192.168.240.1","remote_port":"52898","proto":"HTTP/1.1","method":"POST","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"Sec-Ch-Ua":["\"Chromium\";v=\"103\", \".Not/A)Brand\";v=\"99\""],"Accept":["application/grpc-web-text"],"Accept-Language":["en-US,en;q=0.9"],"X-Forwarded-For":["192.168.240.1"],"Sec-Fetch-Mode":["cors"],"Origin":["http://localhost:9000"],"X-Forwarded-Proto":["http"],"Content-Length":["152"],"Sec-Ch-Ua-Mobile":["?0"],"Accept-Encoding":["gzip, deflate, br"],"Sec-Fetch-Dest":["empty"],"X-Forwarded-Host":["localhost:5052"],"Sec-Ch-Ua-Platform":["\"Linux\""],"X-Grpc-Web":["1"],"Content-Type":["application/grpc-web-text"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Sec-Fetch-Site":["same-site"],"Referer":["http://localhost:9000/"]}},"error":"http2: Transport: cannot retry err [stream error: stream ID 3; PROTOCOL_ERROR; received from peer] after Request.Body was written; define Request.GetBody to avoid this error"}
proxy    | {"level":"error","ts":1657872668.2152588,"logger":"http.log.error","msg":"http2: Transport: cannot retry err [stream error: stream ID 3; PROTOCOL_ERROR; received from peer] after Request.Body was written; define Request.GetBody to avoid this error","request":{"remote_ip":"192.168.240.1","remote_port":"52898","proto":"HTTP/1.1","method":"POST","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"Accept-Encoding":["gzip, deflate, br"],"Sec-Ch-Ua":["\"Chromium\";v=\"103\", \".Not/A)Brand\";v=\"99\""],"Sec-Ch-Ua-Mobile":["?0"],"Sec-Ch-Ua-Platform":["\"Linux\""],"Sec-Fetch-Mode":["cors"],"Connection":["keep-alive"],"Accept":["application/grpc-web-text"],"X-Grpc-Web":["1"],"Content-Type":["application/grpc-web-text"],"Origin":["http://localhost:9000"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Content-Length":["152"],"Sec-Fetch-Site":["same-site"],"Sec-Fetch-Dest":["empty"],"Referer":["http://localhost:9000/"],"Accept-Language":["en-US,en;q=0.9"]}},"duration":0.003161225,"status":502,"err_id":"07zght3j9","err_trace":"reverseproxy.statusError (reverseproxy.go:1184)"}
```

Any new ideas? :thinking:

**Comment by tetra12** [comment]
When I try to do this: 

```bash
http://localhost:5052 {
  reverse_proxy grpc://192.168.1.93:50051    
}
```
I'm getting the following :disappointed: : `"error":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x00\\x00\\x06\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x05\\x00\\x00@\\x00\""}`

The full log :roll_eyes: :
```bash
proxy    | {"level":"debug","ts":1657873059.3257077,"logger":"http.handlers.reverse_proxy","msg":"selected upstream","dial":"192.168.1.93:50051","total_upstreams":1}
proxy    | {"level":"debug","ts":1657873059.3322625,"logger":"http.handlers.reverse_proxy","msg":"upstream roundtrip","upstream":"192.168.1.93:50051","duration":0.004584367,"request":{"remote_ip":"192.168.240.1","remote_port":"52900","proto":"HTTP/1.1","method":"OPTIONS","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"Sec-Fetch-Mode":["cors"],"X-Forwarded-Host":["localhost:5052"],"Accept-Language":["en-US,en;q=0.9"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Accept":["*/*"],"Origin":["http://localhost:9000"],"Sec-Fetch-Site":["same-site"],"X-Forwarded-Proto":["http"],"Access-Control-Request-Method":["POST"],"Accept-Encoding":["gzip, deflate, br"],"Sec-Fetch-Dest":["empty"],"Referer":["http://localhost:9000/"],"Access-Control-Request-Headers":["content-type,x-grpc-web"],"X-Forwarded-For":["192.168.240.1"]}},"error":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x00\\x00\\x06\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x05\\x00\\x00@\\x00\""}
proxy    | {"level":"error","ts":1657873059.3331525,"logger":"http.log.error","msg":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x00\\x00\\x06\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x05\\x00\\x00@\\x00\"","request":{"remote_ip":"192.168.240.1","remote_port":"52900","proto":"HTTP/1.1","method":"OPTIONS","host":"localhost:5052","uri":"/cc_service.CloudControlService/UnaryCall","headers":{"Origin":["http://localhost:9000"],"Sec-Fetch-Site":["same-site"],"Connection":["keep-alive"],"Access-Control-Request-Method":["POST"],"Accept-Encoding":["gzip, deflate, br"],"Accept-Language":["en-US,en;q=0.9"],"Accept":["*/*"],"Access-Control-Request-Headers":["content-type,x-grpc-web"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.53 Safari/537.36"],"Sec-Fetch-Mode":["cors"],"Sec-Fetch-Dest":["empty"],"Referer":["http://localhost:9000/"]}},"duration":0.017830125,"status":502,"err_id":"f09ygcrr4","err_trace":"reverseproxy.statusError (reverseproxy.go:1184)"}
```

@mholt @mohammed90 @francislavoie Any thoughts ? :thinking: :thought_balloon:

**Comment by mohammed90** [comment]
@tetra12 please move this to the forum https://caddy.community/

**Comment by mholt** [comment]
@tetra12 I decided to just hash out a super-sloppy patch in #4896. I have no idea if this works... so could you please try it out? As I have not tested it.

**Comment by mholt** [comment]
There's a lot in this thread, and I'm not sure but I suspect that the https://github.com/mholt/caddy-grpc-web plugin could be helpful here. It can eliminate the need for Envoy as it becomes a grpc-web-to-grpc bridge. @tetra12 and I have both confirmed it works.

Due to inactivity on this thread, I will close it as I suspect it is resolved now. If there's still problems even with the latest version of Caddy (from master, just to be safe) and that plugin, then let us know in a new issue! Thanks!
